### PR TITLE
1062 direct delegation

### LINKED ISSUE #1062: Direct delegation
Responsibles should see a button next to each course on their index page that allows them to delegate the preparation of this course to another user. This button should open a selection field (e.g. in a modal) where the user can be specified. Submitting this would then add the selected user as an editor to the course and optionally send a message to the user, informing about the delegation.

### PR DESCRIPTION
**Pull request by He3lixxx**
fixes #1062

### CODE REVIEW

File: evap/contributor/templates/contributor_index.html [hunk 1]
```diff
@@ -103,6 +106,14 @@
     {% endfor %}
 {% endblock %}
 
+{% block modals %}
+    {{ block.super }}
+    {% trans 'Delegate the preparation' as title %}
+    {% trans 'Do you really want to delegate the preparation of the course <strong id="data-label"></strong>?' as question %}
+    {% trans 'Delegate preparation' as action_text %}
+    {% include 'user_profile_select_modal.html' with modal_id='addEditorModal' title=title form=delegate_selection_form question=question action_text=action_text btn_type='primary' %}
+{% endblock %}
```
Comment (karyon) [inline evap/contributor/templates/contributor_index.html]: unless you plan to include that modal in your next PR a second time, i'd say inline all of that. maybe keep the modal_id as a variable, but the rest, i'd inline it.

File: evap/contributor/views.py [hunk 2]
```diff
@@ -162,3 +165,32 @@ def course_preview(request, course_id):
...
+    body_params = subject_params
+
+    # we don't provide the request here since send_to_user only uses it to display a warning message in case the user does not have
+    # an email address. In this special case, we don't want that warning. Instead, we want a mail to the admins.
+    EmailTemplate.send_to_user(delegate_user, template, subject_params, body_params, use_cc=True, additional_cc_user=request.user)
+
+    messages.add_message(
+        request,
+        messages.SUCCESS,
+        "User {} was added as a contributor for course {} and sent an email with further information.".format(str(delegate_user), str(course))
```
Comment (karyon) [inline evap/contributor/views.py]: `and was sent`, otherwise it reads `User ... sent an email`

File: evap/contributor/tests/test_views.py [hunk 3]
```diff
@@ -1,11 +1,61 @@
...
+    def setUpTestData(cls):
+        cls.course = mommy.make(Course, state='prepared')
+
+        cls.responsible = mommy.make(UserProfile)
+        cls.non_responsible = mommy.make(UserProfile, email="a@b.c")
+        mommy.make(Contribution, course=cls.course, contributor=cls.responsible, can_edit=True, responsible=True, comment_visibility=Contribution.ALL_COMMENTS)
+
+    def test_direct_delegation_request(self):
+        data = {"delegate_user": self.non_responsible.id}
+        page = self.app.post(reverse('contributor:course_direct_delegate', args=[self.course.id]), params=data, user=self.responsible).follow()
```
Comment (karyon) [inline evap/contributor/tests/test_views.py]: ...maybe we should decide whether we want to use that reverse or not, and then do that everywhere.

File: evap/contributor/views.py [hunk 4]
```diff
@@ -162,3 +165,32 @@ def course_preview(request, course_id):
...
+@editor_or_delegate_required
+def direct_delegate_course_editing_to_user(request, course_id):
+    delegate_user_id = request.POST.get("delegate_user")
+
+    course = get_object_or_404(Course, id=course_id)
+    delegate_user = get_object_or_404(UserProfile, id=delegate_user_id)
+
+    contribution, __ = Contribution.objects.get_or_create(course=course, contributor=delegate_user)
+    contribution.can_edit = True
+    contribution.save()
```
Comment (karyon) [inline evap/contributor/views.py]: check out update_or_create, maybe that better expresses what you're doing.

File: evap/evaluation/migrations/0079_create_direct_delegation_email_template.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('evaluation', '0078_populate_voter_and_participant_count'),
+    ]
+
+    operations = [
+        migrations.RunPython(insert_emailtemplates, reverse_code=remove_emailtemplates),
+    ]
```
Comment (karyon) [inline evap/evaluation/migrations/0079_create_direct_delegation_email_template.py]: right now there are two migration files in this  PR.

File: evap/contributor/tests/test_views.py [hunk 6]
```diff
@@ -14,6 +64,34 @@ class TestContributorView(ViewTest):
...
+
+        user1 = mommy.make(UserProfile)
+        user2 = mommy.make(UserProfile, is_active=False)
+
+        responsible = mommy.make(UserProfile)
+        mommy.make(Contribution, course=course, contributor=responsible, can_edit=True, responsible=True, comment_visibility=Contribution.ALL_COMMENTS)
+
+        page = self.app.get('/contributor/', user=responsible, status=200)
+        self.assertContains(page, str(user1))
+        self.assertNotContains(page, str(user2))
```
Comment (karyon) [inline evap/contributor/tests/test_views.py]: i feel these two tests test just a few characters of code each which have very obvious functionality. i think they provide rather little value.

File: evap/contributor/templates/contributor_index.html [hunk 7]
```diff
@@ -118,6 +145,55 @@
...
+            </div>
+        </div>
+
+        <script type="text/javascript">
+            function {{ modal_id }}Show(courseName, action) {
+                // set form's action location
+                $('#{{ modal_id }} form').attr('action', action);
+
+                // show modal
+                $('#{{ modal_id }}').modal();
```
Comment (janno42) [inline evap/contributor/templates/contributor_index.html]: shouldn't this be the last operation?

File: evap/evaluation/migrations/0081_create_direct_delegation_email_template.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
+# Generated by Django 2.0.7 on 2018-08-13 21:09
+
+from django.db import migrations
+
+TEMPLATE_NAME = "Direct Delegation"
+TEMPLATE_SUBJECT = "[EvaP] Bitte Evaluierung vorbereiten / Please prepare evaluation"
+
+
+def insert_emailtemplates(apps, _schema_editor):
```
Comment (janno42) [inline evap/evaluation/migrations/0081_create_direct_delegation_email_template.py]: singular ;)

File: evap/evaluation/migrations/0081_create_direct_delegation_email_template.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    EmailTemplate = apps.get_model("evaluation", "EmailTemplate")
+    EmailTemplate.objects.create(name=TEMPLATE_NAME, subject=TEMPLATE_SUBJECT, body="")
+
+
+def remove_emailtemplates(apps, _schema_editor):
```
Comment (janno42) [inline evap/evaluation/migrations/0081_create_direct_delegation_email_template.py]: singular

File: evap/contributor/templates/contributor_index.html [hunk 10]
```diff
@@ -114,7 +114,7 @@
                                             {% if not course|has_nonresponsible_editor %}
                                                 <a href="#" class="btn btn-sm btn-dark" data-toggle="tooltip"
                                                     data-placement="top" title="{% trans 'Delegate preparation' %}"
-                                                    onclick="addEditorModalShow(`{{ course.name }}`, `{% url 'contributor:course_direct_delegation' course.id %}`);return false;"
+                                                    onclick="delegateSectionModalShow(`{{ course.name }}`, `{% url 'contributor:course_direct_delegation' course.id %}`);return false;"
```
Comment (janno42) [inline evap/contributor/templates/contributor_index.html]: i asked for `delegateSelectionModal` :D

File: evap/contributor/views.py [hunk 11]
```diff
@@ -196,7 +196,7 @@ def course_direct_delegation(request, course_id):
     messages.add_message(
         request,
         messages.SUCCESS,
-        "User {} was added as a contributor for course {} and was sent an email with further information.".format(str(delegate_user), str(course))
+        "{} was added as a contributor for course {} and was sent an email with further information.".format(str(delegate_user), str(course))
```
Comment (janno42) [inline evap/contributor/views.py]: please put some `"` around the course name

**Comment by He3lixxx** [comment]
Sorry, seems like I was distracted. Fixed these.

File: evap/contributor/tests/test_views.py [hunk 12]
```diff
@@ -25,7 +25,7 @@ def test_direct_delegation_request(self):
 
         self.assertContains(
             page,
-            "User {} was added as a contributor for course {} and was sent an email with further information.".format(str(self.non_responsible), str(self.course))
+            "{} was added as a contributor for course {} and was sent an email with further information.".format(str(self.non_responsible), str(self.course))
```
Comment (janno42) [inline evap/contributor/tests/test_views.py]: got distracted again? ;) you didn't put the `"`s here, now the tests are failing.

**Comment by janno42** [comment]
one last thing: can you please fix the import order codacy complains about?

**Pull request merged**